Migrate detailed guide document

sairamya93 opened the pull request:
, [Jira issue PNP-5856](https://gov-uk.atlassian.net/browse/PNP-5856)⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

To handle the rendering of `detailed_guide` document type from government_frontend to frontend as part of [RFC 175](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-175-frontend-fewer-apps.md)

## Why

[Trello card](https://trello.com/c/8HsfHbJ6)

## How

To add a scoped route for, a controller, a model and a view. Here we refactor the methods in the original `detailed_guide` presenter into the appropriate model, presenter and its related concerns required to get detailed_document working as expected.

## Detailed guide example pages

- Page with contents list: /guidance/guidance-on-devolution
- Guidance with logo: /guidance/england-2014-to-2020-european-structural-and-investment-funds
- Withdrawn guidance page: /guidance/eu-rules-on-the-use-of-chemicals 
- Translation page: /guidance/prepare-a-charity-annual-return.cy

## Screenshots?

| Government-frontend | Frontend |
|-------|--------|
|![image](https://github.com/user-attachments/assets/020c6aa4-552a-45ae-9631-9ea63a21452a)|![image](https://github.com/user-attachments/assets/9d8032fd-b69b-4ab6-bc0f-b3aae980a5b0)|
|![image](https://github.com/user-attachments/assets/194a57ed-4867-4829-8bdb-cde9d0326b77)|![image](https://github.com/user-attachments/assets/82aa3204-ea56-4a07-ac90-d3f6e5ce7dbc)|
|![image](https://github.com/user-attachments/assets/d955289e-ef89-4319-a44b-c94c09bb1b64)|![image](https://github.com/user-attachments/assets/7e6572fa-3b67-4de3-91b8-72122c2ddb1f)|
|![image](https://github.com/user-attachments/assets/bb7d5cdb-88e0-4710-a41a-7c8f2c817836)|![image](https://github.com/user-attachments/assets/13860360-c798-4fb1-a6fe-e07637d60b47)|
|![image](https://github.com/user-attachments/assets/43dcf0d7-e885-4045-a728-f4b62589d8fa)|![image](https://github.com/user-attachments/assets/ccb4d39d-d76e-4a29-8256-7948d2a2e036)|
|![image](https://github.com/user-attachments/assets/d53555bd-fe8c-4580-b658-6a1465910062)|![image](https://github.com/user-attachments/assets/4ceaa956-dae0-42f7-8e66-79a745087c11)|
|![image](https://github.com/user-attachments/assets/ce78c2dc-4190-424b-a9c2-8befacd5c6dc)|![image](https://github.com/user-attachments/assets/354d18ba-0733-4b1f-9a74-eaf38ff0d8b6)|
|![image](https://github.com/user-attachments/assets/29a90b14-5fc0-4344-9816-aab1aace9ed2)|![image](https://github.com/user-attachments/assets/7c8613f1-f285-41ec-b364-8a9821a20a9a)|

sairamya93 committed:
Update README.md and govuk_examples.yml

sairamya93 committed:
Modify detailed_guide view

- Added the title to also handle pages that are withdrawn
- Updated the metadata in the header to fetch from govuk_publishing_components directly
- Added the description tag in the header

sairamya93 committed:
Add presenter for detailed_guide

sairamya93 committed:
Modify view - update with heading component

- Updated the code to retrieve 'title' value directly instead of the method 'title_and_context'
- Replaced title with heading component and added '.name' to retrieve  pluralisation values for context keys

sairamya93 committed:
Add request tests for detailed guide

sairamya93 committed:
Add detailed guide system tests

Audit trail: https://github.com/alphagov/government-frontend/blob/e33f1cdf2821faea6a2d75041cd6b6b96a1cdce8/test/integration/detailed_guide_test.rb

sairamya93 committed:
Modify controller to load presenter object

 - Add cacheable in controller

sairamya93 committed:
Modify DetailedGuide model - Retrieve headers

- Added method to retrieve headers from content_item similar to the changes made in corporate information pages.This headers will be formatted in the presenter to display as contents list.
Ref. link: https://github.com/alphagov/frontend/pull/4608/commits/96811ffd0232c2778d605cf616111ad552c4d610

sairamya93 committed:
Add logo in detailed_guide model

- As logo is required to be displayed along with metadata for detailed_guide, I have copied over the logo part.
Example: https://www.gov.uk/guidance/england-2014-to-2020-european-structural-and-investment-funds
- Copied over logo method from government-frontend for displaying the logo in detailed_guide page
Audit trail: https://github.com/alphagov/government-frontend/blob/a7357e18c732e454dab89b56b1e33d99fff7f428/app/presenters/detailed_guide_presenter.rb
Audit trail for test: https://github.com/alphagov/government-frontend/blob/a7357e18c732e454dab89b56b1e33d99fff7f428/test/presenters/detailed_guide_presenter_test.rb#L77

- Copied over metadata logo part and related metadata-logo class stylesheets from government-frontend to display the logo along with the metadata in the page
Audit trail: https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/views/shared/_publisher_metadata_with_logo.html.erb#L10-L16
Audit trail for CSS: https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/assets/stylesheets/helpers/_publisher-metadata-with-logo.scss#L15-L24

sairamya93 committed:
Add route and controller for detailed_guide

sairamya93 committed:
Add View for detailed_guide

- Copied the View from government-frontend
- Audit trail: https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/views/content_items/detailed_guide.html.erb

sairamya93 committed:
Update View tests for publisher dates with notification button

sairamya93 committed:
Add detailed_guide locales

- The below command is run based on the PR[https://github.com/alphagov/frontend/pull/4848] to include .name for the pluralisation keys copied from government-frontend
- Ran the command: govuk-docker-run rake "consolidation:copy_translation[content_item.schema_name.detailed_guide,formats.detailed_guide.name]"
- Audit trail: https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/config/locales/en.yml#L139

deborahchua reviewed (changes requested):
This is coming along nicely! 

I've left a few comments as part of the initial review, mainly about the order of commits and updating some of the commit messages.

sairamya93 committed:
Add related concerns to Detailed guide model

- Added Emphasised organisations concern,contributors method and tests as handled in the migration of routes for case studies and speeches for rendering organisations as part of publisher-metadata values in the page.
Audit trail: https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/presenters/detailed_guide_presenter.rb#L4
Ref. for the change in the organisation concern: https://github.com/alphagov/frontend/pull/4653/commits/95c0852a6c1047e7bc827c2a8c0ff416bc3c15fe

- Added Updatable concern to retrieve published_dates values from  shared partial - published_dates_with_notification_button
- Added Political concern as in government-frontend to handle historically_political values in view
Audit trail: https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/presenters/detailed_guide_presenter.rb

- Added withdrawn tests as the Withdrawn concern is already added in the ContentItem model - https://github.com/alphagov/frontend/blob/8f4326c056248a0f6a1b90b69c61ce3ff9636a03/app/models/content_item.rb#L4

KludgeKML commented on app/views/detailed_guide/show.html.erb:
```diff
@@ -57,13 +57,13 @@
-      <%= render 'govuk_publishing_components/components/govspeak', {} do %>
-        <%= raw(@content_item.govspeak_body[:content]) %>
+      <%= render "govuk_publishing_components/components/govspeak", {} do %>
+        <%= raw(content_item.body.html_safe) %>
```
There's a bunch of lint fixing in this commit as well as the line mentioned in the description. (you could either update the description of the commit, or better yet break it into two commits - one for linting, one for the change)

KludgeKML commented on spec/support/devolved_nations_helper.rb:
```diff
@@ -0,0 +1,10 @@
+module DevolvedNationsHelpers
+  def has_devolved_nations_component(text, nations = nil)
```
You could write this as a full matcher, allowing you to do stuff in your tests like:

`expect(page).to have_devolved_nations_component("Applies to England")`

KludgeKML reviewed (changes requested):
Lookgin good! But a few changes needed. In general. could you follow Leena's request for adding a "why" to the commit messages, especially if they're on the large side.

sairamya93 commented on app/views/detailed_guide/show.html.erb:
```diff
@@ -25,12 +25,24 @@
+    <% withdrawn_time_tag = tag.time(display_date(content_item.withdrawn_at), datetime: content_item.withdrawn_at) %>
+
+    <%= render "govuk_publishing_components/components/notice", {
+      title: I18n.t("withdrawn_notice.title", schema_name: I18n.t("formats.#{content_item.schema_name}", count: 1, locale: :en).downcase, withdrawn_time: withdrawn_time_tag).html_safe,
```
That's right. When I ran the tests, the withdrawn test was failing and I had added `.name` in but I think this particular view is outdated maybe.
https://github.com/alphagov/frontend/blob/4d4e5ca30c666769d5cc32f1d09f7dd0d3db5ece/app/views/detailed_guide/show.html.erb#L40

sairamya93 commented on app/presenters/detailed_guide_presenter.rb:
```diff
@@ -1,2 +1,26 @@
 class DetailedGuidePresenter < ContentItemPresenter
+  def headers_for_contents_list_component
```
Thanks! Exactly,I had the same thought when I was working on it but didn't get around to discuss on that

sairamya93 commented on spec/models/detailed_guide_spec.rb:
```diff
@@ -3,4 +3,24 @@
+  it_behaves_like "it has historical government information", "detailed_guide", "political_detailed_guide"
+  it_behaves_like "it has updates", "detailed_guide", "detailed_guide"
+  it_behaves_like "it has no updates", "detailed_guide", "national_applicability_detailed_guide"
+  it_behaves_like "it can be withdrawn", "detailed_guide", "withdrawn_detailed_guide"
```
Ah yes we have the `Withdrawable` concern added in the [`ContentItem` model](https://github.com/alphagov/frontend/blob/8f4326c056248a0f6a1b90b69c61ce3ff9636a03/app/models/content_item.rb#L4) itself and so we don't explicitly add it to individual document models

sairamya93 committed:
Update Detailed guide presenter

- As implemented in corporate information pages, we exclude H3 and below headings and only include H2 headings extracted from content_item headers attribute fordisplaying the detailed guide contents list.
Reference link: https://github.com/alphagov/frontend/pull/4608/commits/8cc16bc04299bf59efc037f9ecf889145cf231e6
Audit trail for ContentsList in government-frontend: https://github.com/alphagov/government-frontend/blob/a7357e18c732e454dab89b56b1e33d99fff7f428/app/presenters/content_item/contents_list.rb

sairamya93 committed:
Update View - To modify publisher_metadata,withdrawn component,contents list

- Replaced the shared parital with publisher_metadata in line with the changes introduced here[ad511e0] as part of migration to frontend. The logo part is required for detailed_guide and hence added in the subsequent commit.
- Updated withdrawn component as for the other migrated routes with the changes introduced here[4b1d2d4]
- Modified contents list to retrieve the h2 headers from presenter as implemented in corporate information pages[96811ff]

sairamya93 commented on app/views/detailed_guide/show.html.erb:
```diff
@@ -11,8 +11,8 @@
       text: content_item.title,
-      context: I18n.t("formats.#{content_item.document_type}", count: 1),
-      context_locale: t_locale_fallback("formats.#{content_item.document_type}", count: 1),
+      context: I18n.t("formats.#{content_item.document_type}.name", count: 1),
```
I have included that in this commit 59f1ed7 that updates the heading after updating the locales with `.name`

sairamya93 committed:
Fix linting in View file

Ran the command: govuk-docker-run bundle exec erb_lint --lint-all —autocorrect

sairamya93 committed:
Update View with html_safe for body content

- Updated the govspeak body with html_safe body directly
Audit trail: https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/app/presenters/content_item/body.rb#L9

deborahchua commented on app/views/detailed_guide/show.html.erb:
```diff
@@ -0,0 +1,69 @@
+
+<% content_for :extra_headers do %>
+  <%= render "govuk_publishing_components/components/machine_readable_metadata", { content_item: @content_item.to_h, schema: :faq } %>
+  <meta name="description" content="<%= strip_tags(@content_item.description) %>">
```
We can update all instances of `@content_item` to `content_item` throughout.

sairamya93 committed:
Replace @content_item and fix lint for lead_paragraph

sairamya93 committed:
Replace @content_item with content_item for national_applicability

KludgeKML commented on spec/system/detailed_guide_spec.rb:
```diff
@@ -0,0 +1,138 @@
+
+      expected_tracking_bottom = single_page_notification_button_ga4_tracking(2, "Footer")
+      actual_tracking_bottom = JSON.parse(buttons.last["data-ga4-link"])
+      expect(actual_tracking_bottom).to eq(expected_tracking_bottom)
```
non-blocking nitpick - except where it's the first line in a block or following another expect line, could we always have a space before each expect?

KludgeKML reviewed (approved):
LGTM - one non-blocking nitpick.

deborahchua reviewed (approved):
🚀
